### DISCUSSION TITLE
is it possible to implement multi-process app using Avalonia ?

### DISCUSSION THREAD

**Discussion by ExtremeGTX** [AvaloniaUI/Avalonia#9738] [Q&A] [answered]
Hello AvaloniaUI Team,
Thanks for providing such great framework.

I have a question about implementing an App in an architecture similar to Browsers Edge/Chrome/Firefox where there is a single process which has all Display/UserInteractions but at the same time each sub-process is isolated from the same process.

I checked the issue here https://github.com/AvaloniaUI/Avalonia/issues/2379 to understand that it is not possible for cross frameworks, but what is the case if Avalonia App wants to render it is UI in another Avalonia App on the same OS ?

##### P.S.:
Example on WPF render on Win32 window using custom `HwndSource` [here](https://stuff.seans.com/2008/09/01/writing-a-screen-saver-in-wpf/)

**Comment by kekekeks** [comment, marked as answer]
1) Cross-app control embedding isn't supported on macOS at all
2) Browsers are using shared memory (DXGI surface handle, IOSurface, etc) to send rendered frames to the GUI process.

**Reply by ExtremeGTX** [reply]
> 1. Cross-app control embedding isn't supported on macOS at all

Do you mean it is supported on Windows/Linux ? if yes could you guide me where could i start ?

> 2. Browsers are using shared memory (DXGI surface handle, IOSurface, etc) to send rendered frames to the GUI process.

To give you some background, i'm checking the feasibility of replacing Qt with .NET+Avalonia on Embedded Linux based system. currently similar functionality is provided by [Qt Application Manager](https://doc.qt.io/QtApplicationManager/introduction.html)

**Reply by kekekeks** [reply]
Avalonia doesn't provide anything out of the box, but it should be technically possible by manually tinkering with libdrm, libgbm, EGLSurface and some IPC based on unix sockets.

The process would look somewhat like this:
1) embedded app creates a GBM surface handle and uses it as a render target, the required code can be found in [DrmOutput.cs](https://github.com/AvaloniaUI/Avalonia/blob/24d447bc45bd9e2e12cf5475feb3970b69f83026/src/Linux/Avalonia.LinuxFramebuffer/Output/DrmOutput.cs#L18)
2) embedded app connects to the host app via a unix socket and uses sendfd call to pass a GBM surface handle
3) host app wraps that GBM handle in a EGLSurface and uses that EGLSurface to render content using OpenGlControlBase (there will be some easier way to use GBM handles as regular bitmaps somewhere in the next few months)


If that looks too complicated and you are happy with CPU-powered rendering, you can also use `RemoteWidget`/`RemoteServer` controls that are utilized by our previewer infrastructure. Those can work even over TCP.

**Reply by ExtremeGTX** [reply]
Thanks a lot for providing this overview. appreciate it.
